Fuzz contract constructor arguments

andremarianiello opened the pull request:
In order to have constructor information, I had to update to hevm-0.21. This required bumping the stack resolver to have a new enough version of the time package that was compatible with hevm, and removing some version constraint (though I didn't try very hard to have a tight bound, I was happy to let stack handle the version compatibilities).

I moved VM initialization from before TestableContract construction to after. Now it happens right before executing each generated sequence of SolCalls, and if the contract reverts, that test is skipped, just like what was being done when a random function call caused a revert.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/crytic/echidna?pullRequest=141) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/crytic/echidna?pullRequest=141) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/crytic/echidna?pullRequest=141) it.</sub>

andremarianiello committed:
Stop using deprecated functions from Data.Yaml

andremarianiello committed:
Update to newest hevm

andremarianiello committed:
Add constructor fuzzing

ggrieco-tob commented on .travis/install-secp256k1.sh:
```diff
@@ -0,0 +1,18 @@
+# don't install shared library so linker must used static
+# enable pic so static library can link against dynamic correctly
+./configure --enable-module-recovery --enable-shared=no --with-pic
+sudo make install
```
README instructions should be updated explain how the installation process should be performed. Keep in mind that we don't want to recommend people to install libraries using `sudo make install` since it's very difficult to un-install them later (for travis scripts, it's ok, we don't care).

andremarianiello replied:
I added a bunch of instructions to the README to cover this. Take a look and let me know if you want me to make any changes.

ggrieco-tob commented on stack.yaml:
```diff
@@ -1,25 +1,28 @@
+- multiset-0.3.4.1
+- text-format-0.3.2
+
+nix:
```
I don't understand why this section section is necessary if the travis installation does not require NIX. Please remove this if it is useless. Also, update the README to make sure the instructions are correct.

andremarianiello replied:
It is not necessary for CI, but it is necessary if someone wants to use stack's nix integration. It let's someone build echidna using `stack build --nix` instead of relying on system libraries that may or may not exist. In the case of libsecp256k1, it almost certainly would not exist. This list of packages also serves as documentation of what system deps echidna needs. I think we should keep this here so I can put `stack build --nix` as one option for building (I will also make sure that the steps for building without nix are documented as well). Does this sound reasonable to you?

andremarianiello replied:
bump

ggrieco-tob replied:
While it sounds reasonable to me, we are not currently supporting nix to build echidna. We would love to do it, but we a have a few reasons to avoid it. If we decided to accept this change, we *should* keep updated this list of packages, in case someone want to use it. But since we are not using nix, that's why I asked you to delete that section. If you want, feel free to support an installation using nix in your own branch. We could even link your branch from our README, in case someone is interested in using nix.

asymmetric replied:
What's the advantage of using `stack --nix` rather than just adding a nix derivation for echidna to nixpkgs?

Also, IIRC all hackage packages are automatically built on nixpkgs.

asymmetric replied:
@andremarianiello once echidna is on hackage (the Haskell package repository) it **should** be picked up within a week or so by the nix infrastructure automatically.

There might be build issues and then we'd have to look into it, but I think pushing to Hackage is the first step.

asymmetric replied:
But now I realize that the above tool only works with packages using Cabal, not Stack. There's [stack2nix](https://github.com/input-output-hk/stack2nix) for that, but the last time I tried packaging echidna with it, it was broken. :man_shrugging:

andremarianiello replied:
@ggrieco-tob Can expand on

> we are not currently supporting nix to build echidna. We would love to do it, but we a have a few reasons to avoid it.

gustavo-grieco reviewed (changes requested):
This is the first part of the review regarding the use of the new HEVM.

ggrieco-tob commented:
Can you add a simple example in the `solidity` folder to quickly test this feature? (we want to have a small solidity contract test suite for all the echidna features in the near future).

ggrieco-tob commented:
It seems that the contract argument fuzzing is not well integrated into the coverage mode. Once echidna tests and runs certain constructor paramters, they are not properly saved into the `outdir` directory (only the function calls are saved). Without this, other echidna instances won't be able to replicate a test case and the "distributed mode" will be quite ineffective. Can you fix this?

andremarianiello commented:
> Can you add a simple example in the `solidity` folder to quickly test this feature? (we want to have a small solidity contract test suite for all the echidna features in the near future).

Done

andremarianiello commented:
> It seems that the contract argument fuzzing is not well integrated into the coverage mode. Once echidna tests and runs certain constructor paramters, they are not properly saved into the `outdir` directory (only the function calls are saved). Without this, other echidna instances won't be able to replicate a test case and the "distributed mode" will be quite ineffective. Can you fix this?

I now store the used arguments in the json that is output to coverage files. I think will allow faithful reproduction of the test case, as requested. I also included the constructor args as part of the "unique key" for the set of tests. If you would rather the CoverageInfo remain the only member of that key, I can revert that change.

ggrieco-tob commented:
Great!, I will review the new changes during this week.

andremarianiello commented:
@ggrieco-tob Any more suggestions?

ggrieco-tob commented:
Sorry for the delay!, I had trouble merging your last commits (maybe because of the forced-pushes?). Anyway, it is fixed now so I'm reviewing your latest changes..

ggrieco-tob commented:
I'm testing your latest commits with a modification of the `solidity/coverage/single.sol` file:

```
contract C {
  bool state = true;

  function C(uint x) {
     require(x <= 1024);
  }

  function f(uint x, uint y, uint z) {
    require(x == 12);
    require(y == 8);
    require(z == 0);
    state = false;
    return; 
  }

  function echidna_state() returns (bool) {
    return state;
  }

}
```

If you run this file with new code, you should get:

```
$ echidna-test single.sol --config test.yaml 
...
"Analyzing contract: single.sol:C"
1000000
Found 0 new testcases!
Found -6175 new testcases!
...
```

It takes a long time to find the a way to make that trivial property false. However, if you use the original test, it founds the counter-example almost immediately. I assuming there is some issues with the new code in the generation/fuzzing of the transactions, once echidna finds a way to avoid the revert in the constructor. Let's fix this before merging.

Also, can you fix the report of the negative numbers in number of test cases? :smile:

andremarianiello commented:
@ggrieco-tob Thanks for taking a look; I'll investigate.

andremarianiello committed:
Add constructor args to output for failed tests

andremarianiello commented:
@ggrieco-tob I was using the constructor args AND the coverage info for identifying unique code paths. This was causing the number of unique test cases to blow up, and was unnecessary. I changed it back to using only the coverage info. Also, I commit the change you made to the single.sol contract.

I could not reproduce the negative test cases output you got, so you might have to show me how you did that.

I also added constructor args to the output for failed properties.

ggrieco-tob commented:
The issue with the negative number of test cases is fixed! However, I'm testing your latest commits with a modification of the `solidity/coverage/multi.sol` file and I think I found a regression:

```
contract C {
  bool state1 = false;
  bool state2 = false;
  bool state3 = false;
  
  function f(uint x) {
    require(x == 12);
    state1 = true;
  }

  function g(uint y) {
    require(state1);
    require(y == 8);
    state2 = true;    
  }
  
  function h(uint z) {
    require(state2);
    require(z == 0);
    state3 = true;
  }
  
  function ul(uint) {
    uint x = 0;
    return;
  }
  
  function echidna_state3() returns (bool) {
    return (!state3);
  }

}
```

To run this example:

```
$ mkdir -p outdir
$ rm outdir/* 
$ echidna-test multi.sol --config test.yaml
```

Using the original `dev-no-hedgehog` branch, you get:

```
"Analyzing contract: multi.sol:C"
1000000
Found 0 new testcases!
Found 0 new testcases!
Failed property "echidna_state3"
Original input:
h(0) by a329c0648769a73afac7f9381e08fb43dbea70
g(8) by a329c0648769a73afac7f9381e08fb43dbea70
ul(69886639306845688416592989777230678738175469504079973247) by a329c0648769a73afac7f9381e08fb43dbea70
f(12) by a329c0648769a73afac7f9381e08fb43dbea70
Original events:
[]
Reduced input:
h(0) by a329c0648769a73afac7f9381e08fb43dbea70
g(8) by a329c0648769a73afac7f9381e08fb43dbea70
f(12) by a329c0648769a73afac7f9381e08fb43dbea70
Reduced events:
[]
```

but using your PR:

```
"Analyzing contract: multi.sol:C"
1000000
Found 0 new testcases!
Found 0 new testcases!
Failed property "echidna_state3"
Constructor Arguments:
Original input:
h(23192810032488445957108697628195474228326257) by a329c0648769a73afac7f9381e08fb43dbea70
g(8) by a329c0648769a73afac7f9381e08fb43dbea70
f(12) by a329c0648769a73afac7f9381e08fb43dbea70
Original events:
[]
Reduced input:
h(23192810032488445957108697628195474228326257) by a329c0648769a73afac7f9381e08fb43dbea70
g(8) by a329c0648769a73afac7f9381e08fb43dbea70
f(12) by a329c0648769a73afac7f9381e08fb43dbea70
Reduced events:
[]
```

which is clearly not the correct result. Can you fix this regression?

andremarianiello committed:
Fix passing of call sequence which is minimized and reported

andremarianiello commented:
Good catch. The original call sequence `cs` was being reported, not the mutated call sequence `cs'` which was actually triggering the property failure. Pushed the fix.

ggrieco-tob commented:
Great work!, the bug seems to be fixed. A few small remaining issues: If you create a simple contract like this one:

```
contract C {
    function C(uint x) {
        assert(x == 0);
    }
    function echidna_fail() returns (bool) {
        return false;
    } 
}
```

Then echidna will complain that there are no functions to execute (which is no longer the case in this branch).

Additionally, can you print the constructor call using the same convention as the calls when we print the original and the minimized ones? For instance:

```
Original input:
C(8994384938493) by a329c0648769a73afac7f9381e08fb43dbea70
f(0) by a329c0648769a73afac7f9381e08fb43dbea70
Original events:
[]
Reduced input:
C(42) by a329c0648769a73afac7f9381e08fb43dbea70
f(0) by a329c0648769a73afac7f9381e08fb43dbea70
Reduced events:
[]
```

Also, please verify that the constructor can be minimized and add one example of a small contract in the solidity folder.

Finally, do not forget to fix the warnings, otherwise, the CI tests will fail.

japesinator commented:
Closing, outdated
